third_party/nvidia/lib/Dialect/NVWS/Transforms/LowerAref.cpp

```diff
@@ -55,294 +56,587 @@ namespace triton {
 
 struct ArefValue {
   Value emptyMbars;
   Value fullMbars;
   int depth;
-  SmallVector<ArefUseNode *> users;
-  SmallVector<Value> updatedValues;
-};
-
-struct ArefUseGraph {
-  llvm::MapVector<Operation *, ArefUseNode> nodes;
-  llvm::MapVector<Value, ArefValue> arefs;
+  SmallVector<Value> buffers;
 };
 
-MemDescType getMemDesc(PatternRewriter &rewriter, llvm::ArrayRef<int64_t> shape,
-                       Type intType) {
+MemDescType getBarrierMemDesc(PatternRewriter &rewriter,
+                              llvm::ArrayRef<int64_t> shape) {
   auto ctx = rewriter.getContext();
   Attribute sharedMemorySpace = triton::gpu::SharedMemorySpaceAttr::get(ctx);
   auto barrierCTALayout = CTALayoutAttr::get(
       /*context=*/ctx, /*CTAsPerCGA=*/{1},
       /*CTASplitNum=*/{1}, /*CTAOrder=*/{0});
   auto barrierEncoding =
       SwizzledSharedEncodingAttr::get(ctx, 1, 1, 1, {0}, barrierCTALayout);
-  return MemDescType::get(shape, intType, barrierEncoding, sharedMemorySpace,
+  return MemDescType::get(shape, rewriter.getI64Type(), barrierEncoding,
+                          sharedMemorySpace,
                           /*mutableMemory=*/true);
 }
-MemDescType getBarrierMemDesc(PatternRewriter &rewriter,
-                              llvm::ArrayRef<int64_t> shape) {
-  return getMemDesc(rewriter, shape, rewriter.getI64Type());
-}
 
-Value getBarrierAt(PatternRewriter &rewriter, Location loc, Value mbars,
-                   Value idx) {
+Value getBarrier(PatternRewriter &rewriter, Location loc, Value mbars,
+                 Value idx) {
   auto memDesc = getBarrierMemDesc(rewriter, {1});
   return rewriter.create<triton::gpu::MemDescSubviewOp>(
       loc, memDesc, mbars, SmallVector<Value>{idx});
 }
 
-Value getBarrier(PatternRewriter &rewriter, Location loc, Value mBars,
-                 Value idx) {
-  return getBarrierAt(rewriter, loc, mBars, idx);
+Value getEmptyBarrier(PatternRewriter &rewriter, Location loc, ArefValue aref,
+                      Value arefIdx) {
+  // stage= arefIdx % depth
+  Value stage;
+  auto remsi = rewriter.create<arith::RemSIOp>(
+      loc, arefIdx, rewriter.create<arith::ConstantIntOp>(loc, aref.depth, 32));
+  remsi->setAttr("empty_mbar", rewriter.getUnitAttr());
+  stage = remsi;
+
+  return getBarrier(rewriter, loc, aref.emptyMbars, stage);
 }
 
-void waitOnMbar(PatternRewriter &rewriter, Location loc, Value mBars, Value idx,
-                Value phase) {
-  auto ctx = rewriter.getContext();
-  auto barrier = getBarrier(rewriter, loc, mBars, idx);
-  // wait on empty/full
-  auto waitOp =
-      rewriter.create<triton::nvidia_gpu::WaitBarrierOp>(loc, barrier, phase);
+Value getFullBarrier(PatternRewriter &rewriter, Location loc, ArefValue aref,
+                     Value arefIdx) {
+  Value stage;
+  auto remsi = rewriter.create<arith::RemSIOp>(
+      loc, arefIdx, rewriter.create<arith::ConstantIntOp>(loc, aref.depth, 32));
+  remsi->setAttr("full_mbar", rewriter.getUnitAttr());
```

Comment:
what's the purpose of this attribute?

Reply:
same as above (https://github.com/triton-lang/triton/pull/7479#discussion_r2205987827)